R/transformCounts.R

```diff
@@ -308,88 +303,75 @@ setMethod("relAbundanceCounts",signature = c(x = "SummarizedExperiment"),
              "Try log10 with pseudocount = 1 or other numeric value.",
              call. = FALSE)
     }
-
     mat <- log10(mat)
-
     return(mat)
 }
 
-.calc_pa <- function(mat, threshold){
+.calc_pa <- function(mat, threshold, ...){
     # If value is over zero, gets value 1. If value is zero, gets value 0.
     mat <- (mat > threshold) - 0L
     return(mat)
 }
 
-.calc_hellinger <- function(mat, threshold){
-
+.calc_hellinger <- function(mat, pseudocount, ...){
+    # Gets the relative abundance
+    mat <- .calc_rel_abund(mat)
+    mat <- .apply_pseudocount(mat, pseudocount)
     # If there is negative values, gives an error.
     if (any(mat < 0, na.rm = TRUE)) {
         stop("Abundance table contains negative values and hellinger ",
-             "transformation is being applied without pseudocount. Try ",
-             "hellinger with pseudocount = 1 or other numeric value.",
+             "transformation is being applied without (suitable) pseudocount. ",
+             "Try hellinger with pseudocount = 1 or other numeric value.",
              call. = FALSE)
     }
-
-    # Gets the relative abundance
-    mat <- .calc_rel_abund(mat)
-
     # Takes square root
     mat <- sqrt(mat)
-
     return(mat)
 }
 
 #' @importFrom DelayedMatrixStats colMeans2
-.calc_clr <- function(mat, threshold){
-
+.calc_clr <- function(mat, pseudocount, ...){
+    mat <- .calc_rel_abund(mat)
+    mat <- .apply_pseudocount(mat, pseudocount)
     # If there is negative values, gives an error.
     if (any(mat <= 0, na.rm = TRUE)) {
         stop("Abundance table contains zero or negative values and ",
              "clr-transformation is being applied without pseudocount. ",
              "Try clr with pseudocount = 1 or other numeric value.",
```

Comment:
Same comment here. If pseudocount is added on relative abundances, then 1 is not necessarily a good option(?)

Reply:
Thats up to the user. He has to decide, what to do. A check whether relative abundances are already present is not up to us. (The man page clearly spells out the formula used)

Reply:
What is commonly used for `clr`?

Reply:
Why do we want to recommend using pseudocount 1 here ? - this is potentially misleading.

Reply:
I did assume with Chandlers bug report, that this is usually done like this. First relative abundance, then pseudocount and then the rest of the transformation. 

If that is not the case, than the procedure he used shouldn't work or at least should have the same result.

Reply:
What kind of pseudocount is commonly used for `clr` transformation?

Reply:
The way this should work is that relative abundance is always done before clr transformation (as a preprocessing step). Without pseudocount by default.

Then this process should return the same result regardless on whether it is being applied on relative abundance or count values.

Unfortunately, clr gets easily broken with zero values and most of the time pseudocount is needed. 

If pseudocount is added, the default choice would be to add it on the count data and use pseudocount = 1.

If relative abundances contain zeroes, then it suggests that relabundances were calculated without pseudocount.

If pseudocount is added on relative abundances instead of count data, a default choice could be `pseudocount = min(x[x>0])`.

That should yield identical result (within tolerance) for those two alternative ways to do this in most cases. 

Regarding warning messages, we should either not give recommendations on pseudocount values, or we should give recommendations that remind the user that default choice will depend on the assay type (count / relabundance). 

In general, clr should not be done for other than count or relabundance assay. Perhaps the user should get a warning is such case.

Later when time allows, we should implement rclr which is a robust version that avoids this. I can open issue about that.

Reply:
>If pseudocount is added, the default choice would be to add it on the count data and use pseudocount = 1.
>If relative abundances contain zeroes, then it suggests that relabundances were calculated without pseudocount.
>If pseudocount is added on relative abundances instead of count data, a default choice could be pseudocount = min(x[x>0]).

There is no way, we can check for these things, so it is the user who has to make sure. We can just make sure the calculation is correct, but we cannot make sure that the calculation is the correct choice.

However, we can spell it out even more directly on the man page.

>In general, clr should not be done for other than count or relabundance assay. Perhaps the user should get a warning is such case.

How to check for these cases? I don't see bulletproof way.

Reply:
A PR making sure, that the man page is the best version possible, would be good, but I don't think the code can adjusted to avoid every pitfall there is.

Reply:
Yes there are limits to how much should be done for the user. But:

1) man page should give clear recommendation, explanation, and examples

2) warning messages should not encourage naive users make bad choices (i.e. if default values are recommended for pseudocount, then let us also spell out that different default values are suitable for different cases)